core/src/main/scala/org/apache/spark/scheduler/DecommissionTracker.scala

```diff
@@ -0,0 +1,405 @@
+    // for the Graceful Decommission node.
+    // If the node is already terminated and hostname is reused in that scenario
+    // no need to remove the shuffle entry from map-output tracker
+    if (! decommissionHostnameMap.contains(hostname)) {
+      logInfo(s"""Node $hostname not found in decommisssionTrackerList while """ +
+        """performing shuffle data decommission""")
+      return
+    }
+    // if the terminationTime in the thread is not equal to
+    // terminationTime in decommissionHostnameMap for that
+    // host than Ignore the removeShuffleData
+    if (decommissionHostnameMap(hostname).terminationTime
+      != nodeDecommissionInfo.terminationTime) {
+      logInfo(s"Ignoring removeShuffleData for hostname ${hostname}," +
+        s" since node is already terminated")
+      return
+    }
+
+    // Unregister shuffle data.
+    dagScheduler.map(_.nodeDecommissioned(hostname))
+
+    decommissionHostnameMap(hostname).state = NodeDecommissionState.SHUFFLEDATA_DECOMMISSIONED
+
+    logInfo(s"Node $hostname Shuffle data decommissioned")
+
+    return
+  }
+
+  private def terminate(hostname: String): Unit = {
+    // Not found, only valid scenario is the nodes
+    // has moved back to running state
+    if (!decommissionHostnameMap.contains(hostname)) {
+      logWarning(s"Node $hostname not found in decommisssionTrackerList")
+      return
+    }
+
+    // Remove all the shuffle data of all the executors for the terminated node
+    dagScheduler.map(_.nodeDecommissioned(hostname))
+
+    decommissionHostnameMap(hostname).state = NodeDecommissionState.TERMINATED
+
+    logInfo(s"Node $hostname terminated")
+  }
+
+  def stop (): Unit = {
+    val decommissionNodeCnt = DecommissionTracker.getNodeDecommissionCnt()
+    val fetchFailIgnoreCnt = DecommissionTracker.getFetchFailIgnoreCnt()
+    val fetchFailIgnoreThresholdExceed = DecommissionTracker.getFetchFailIgnoreCntThresholdFlag()
+    val fetchFailedIgnoreThreshold =
+      conf.get(config.GRACEFUL_DECOMMISSION_FETCHFAILED_IGNORE_THRESHOLD)
+    decommissionThread.shutdown()
+    var message = s"Decommission metrics: ${decommissionNodeCnt} nodes decommissioned" +
+      s" (either due to node loss or node block rotation) while running ${conf.getAppId}." +
+      s" Ignored ${fetchFailIgnoreCnt} fetch failed exception caused by decommissioned node."
+
+    if (fetchFailIgnoreThresholdExceed) {
+      message += s"Fetch fail ignore exceeded threshold ${fetchFailedIgnoreThreshold}" +
+        s" causing stage abort"
+    } else {
+      message += s"Fetch fail ignored under allowed threshold ${fetchFailedIgnoreThreshold}"
+    }
+    // logging the metrics related to graceful decommission from decommission tracker
+    logDebug(message)
+  }
+
+}
+
+private[spark] object DecommissionTracker extends Logging {
+  val infiniteTime = Long.MaxValue
+
+  // Stats
+  val decommissionNodeCnt = new AtomicInteger(0)
+  val fetchFailIgnoreCnt = new AtomicInteger(0)
```

Comment:
these are purely for metrics and only logged on stop?  I question is this is really needed or useful then for the added code?

Reply:
These stats which is shown at the end of the application is helpful to find out how the application ran even if there was node decommission event has happened. Instead of checking the entire logs to find out if the node decommission happened or not , is there any fetch failure happened.

Also in future we can some code change where we can add these metrics in spark eventLogs to show in Spark UI these metrics can be used.

These metrics calculated here can be used in multiple purpose for those who wants to send this metrics to there downstream application like signalfx etc. They just have to add the code to pass these stats calculated here to the client.